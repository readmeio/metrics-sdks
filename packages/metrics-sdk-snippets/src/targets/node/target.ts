import type { Target } from '../targets';
import { express } from './express/client';

export const node: Target = {
  info: {
    key: 'node',
    title: 'Node.js',
    extname: '.js',
    default: 'native',
    cli: 'node %s',
  },
  clientsById: {
    express,
  },
};
