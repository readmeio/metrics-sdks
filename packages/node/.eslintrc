{
  "extends": "@readme/eslint-config",
  "root": true,
  "rules": {
    "no-underscore-dangle": ["error", {
      "allow": ["_body"]
    }]
  }
}
