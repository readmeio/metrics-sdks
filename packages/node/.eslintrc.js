module.exports = {
  extends: ['@readme/eslint-config', '@readme/eslint-config/typescript'],
  plugins: ['node', '@typescript-eslint', 'import'],
  root: true,
  rules: {
    // Disable requiring return types because it's too easy to broaden them by accident.
    '@typescript-eslint/explicit-module-boundary-types': 'off',
    '@typescript-eslint/no-empty-function': 'warn',
    '@typescript-eslint/no-var-requires': 'warn',

    'import/no-unresolved': [
      'error',
      {
        ignore: [
          // We're just importing types, so we don't need this unresolved.
          'har-format',
        ],
      },
    ],

    'no-underscore-dangle': ['error', { allow: ['_id', '_body', '__bodyCache'] }],

    // We use `lodash` because it allows for more flexibility that we can't get with standard object accessors.
    'you-dont-need-lodash-underscore/get': 'off',
    'you-dont-need-lodash-underscore/omit': 'off',
  },
  overrides: [
    {
      files: ['examples/**/*.js'],
      rules: {
        // This was getting really weird in the examples/ folder
        // where I know `readmeio` isn't going to be resolvable yet.
        // In some versions of node that is true, but in others it's false
        // which was causing this rule to trigger ðŸ¤¯
        'eslint-comments/no-unused-disable': 'off',
        // We don't want the top level eslint task to fail if the examples
        // folders are missing dependencies. These will be tested separately
        // during the integration testing step.
        'import/no-unresolved': 'off',
      },
    },
  ],
};
