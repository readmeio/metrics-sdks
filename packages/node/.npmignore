.nyc_output/
coverage/
examples/
test/
.eslint*
.prettier*
