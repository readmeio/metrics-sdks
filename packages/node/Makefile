.PHONY: help test
API_KEY ?= $(shell bash -c 'read -p "ReadMe API Key: " api_key; echo $$api_key')

build: ## Build the library dist
	npm run build

install: ## Install all dependencies
	npm ci

lint: ## Run code standard checks
	npx eslint .

lint-fix: ## Attempt to automatically fix any code standard violations
	npx eslint --fix .

serve-metrics-express: ## Start the local Express server to test Metrics
	README_API_KEY=$(API_KEY) npm run --prefix=examples/express start:metrics

serve-metrics-fastify: ## Start the local Fastify server to test Metrics
	README_API_KEY=$(API_KEY) npm run --prefix=examples/fastify start:metrics

serve-metrics-hapi: ## Start the local hapi server to test Metrics
	README_API_KEY=$(API_KEY) npm run --prefix=examples/hapi start:metrics

serve-webhooks-express: ## Start the local Express erver to test webhooks
	README_API_KEY=$(API_KEY) npm run --prefix=examples/express start:webhooks

test: ## Run unit tests
	npm run test

help: ## Display this help screen
	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
