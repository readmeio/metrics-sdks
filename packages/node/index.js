const request = require('r2');
const config = require('./config');

const constructPayload = require('./lib/construct-payload');
const createJWTLink = require('./lib/create-jwt-link');
const getReadmeData = require('./lib/get-readme-data');

// We're doing this to buffer up the response body
// so we can send it off to the metrics server
// It's unfortunate that this isn't accessible
// natively. This may take up lots of memory on
// big responses, we can make it configurable in future
function patchResponse(res) {
  const { write, end } = res;

  res._body = '';

  res.write = (chunk, encoding, cb) => {
    res._body += chunk;
    write.call(res, chunk, encoding, cb);
  };

  res.end = (chunk, encoding, cb) => {
    // Chunk is optional in res.end
    // http://nodejs.org/dist/latest/docs/api/http.html#http_response_end_data_encoding_callback
    if (chunk) res._body += chunk;
    end.call(res, chunk, encoding, cb);
  };
}

module.exports.metrics = (apiKey, group, options = {}) => {
  if (!apiKey) throw new Error('You must provide your ReadMe API key');
  if (!group) throw new Error('You must provide a grouping function');

  const bufferLength = options.bufferLength || config.bufferLength;
  const encoded = Buffer.from(`${apiKey}:`).toString('base64');
  let queue = [];

  return (req, res, next) => {
    const startedDateTime = new Date();
    patchResponse(res);

    function send() {
      // This should in future become more sophisticated,
      // with flush timeouts and more error checking but
      // this is fine for now
      queue.push(constructPayload(req, res, group, options, { startedDateTime }));
      if (queue.length >= bufferLength) {
        request
          .post(`${config.host}/v1/request`, {
            headers: { authorization: `Basic ${encoded}` },
            json: queue,
          })
          .response.then(() => {
            queue = [];
          });
      }

      cleanup(); // eslint-disable-line no-use-before-define
    }

    function cleanup() {
      res.removeListener('finish', send);
      res.removeListener('error', cleanup);
      res.removeListener('close', cleanup);
    }

    // Add response listeners
    res.once('finish', send);
    res.once('error', cleanup);
    res.once('close', cleanup);

    return next();
  };
};

module.exports.login = (apiKey, getUser, options = {}) => {
  if (!apiKey) throw new Error('You must provide your ReadMe API key');
  if (!getUser) throw new Error('You must provide a function to get the user');

  // Make sure api key is valid
  getReadmeData(apiKey);

  return async (req, res, next) => {
    let u;
    try {
      u = getUser(req);
    } catch (e) {
      // User isn't logged in
    }

    if (!u) {
      const domain = req.headers['x-forwarded-host'] || req.get('host');
      const fullUrl = `${req.protocol}://${domain}${req.originalUrl}`;
      return res.redirect(`${options.loginUrl}?redirect=${encodeURIComponent(fullUrl)}`);
    }

    try {
      const jwtUrl = await createJWTLink(apiKey, u, req.query.redirect);
      return res.redirect(jwtUrl);
    } catch (e) {
      return next(e);
    }
  };
};
