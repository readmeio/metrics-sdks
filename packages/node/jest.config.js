/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
module.exports = {
  coveragePathIgnorePatterns: ['dist/'],
  globals: {
    'ts-jest': {
      tsconfig: '__tests__/tsconfig.json',
    },
  },
  globalSetup: '<rootDir>/__tests__/jest.setup.js',
  modulePathIgnorePatterns: ['<rootDir>/dist'],
  preset: 'ts-jest',
  testEnvironment: 'node',
  testMatch: ['**/__tests__/**/(*.)+(spec|test).[jt]s?(x)', '**/?(*.)+(spec|test).[jt]s?(x)'],
};
