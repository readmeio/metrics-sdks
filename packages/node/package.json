{
  "name": "readmeio",
  "description": "Track your API metrics within ReadMe.",
  "version": "5.1.1",
  "license": "ISC",
  "homepage": "https://readme.com/metrics",
  "bugs": {
    "email": "support@readme.io",
    "url": "https://github.com/readmeio/metrics-sdks/issues"
  },
  "repository": {
    "type": "git",
    "url": "git@github.com:readmeio/metrics-sdks.git"
  },
  "keywords": [
    "api-metrics",
    "readme"
  ],
  "engines": {
    "node": ">=14"
  },
  "main": "dist/src/index.js",
  "types": "dist/src/index.d.ts",
  "files": [
    "dist"
  ],
  "dependencies": {
    "content-type": "^1.0.4",
    "find-cache-dir": "^3.3.1",
    "flat-cache": "^3.0.4",
    "lodash": "^4.17.15",
    "node-fetch": "^2.6.7",
    "timeout-signal": "^1.1.0",
    "uuid": "^8.2.0"
  },
  "devDependencies": {
    "@readme/eslint-config": "^8.7.0",
    "@types/har-format": "^1.2.7",
    "@types/jest": "^27.4.1",
    "@types/multer": "^1.4.7",
    "@types/node": "^17.0.24",
    "body-parser": "^1.19.2",
    "eslint": "^8.10.0",
    "express": "^4.17.1",
    "form-data": "^4.0.0",
    "is-valid-uuid-v4": "^1.2.0",
    "jest": "^27.5.1",
    "multer": "^1.4.4",
    "nock": "^13.2.4",
    "prettier": "^2.5.1",
    "rimraf": "^3.0.2",
    "supertest": "^6.0.1",
    "ts-jest": "^27.1.3",
    "typescript": "^4.6.2"
  },
  "scripts": {
    "build": "tsc",
    "lint": "eslint .",
    "prepack": "npm run build",
    "pretest": "npm run build && npm run lint",
    "prettier": "prettier --list-different --write \"./**/**.{js,ts}\"",
    "test": "jest --coverage"
  },
  "prettier": "@readme/eslint-config/prettier"
}
