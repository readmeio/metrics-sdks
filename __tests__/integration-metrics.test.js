import 'isomorphic-fetch';
import { spawn } from 'node:child_process';
import { once } from 'node:events';
import fs from 'node:fs/promises';
import http from 'node:http';
import net from 'node:net';
import { cwd } from 'node:process';
import { Readable, Transform } from 'node:stream';

import chai, { expect } from 'chai';
import { FormDataEncoder } from 'form-data-encoder';
import { File, FormData } from 'formdata-node';
import getPort from 'get-port';

import chaiPlugins from './helpers/chai-plugins.js';

if (!process.env.EXAMPLE_SERVER) {
  // eslint-disable-next-line no-console
  console.error('Missing `EXAMPLE_SERVER` environment variable');
  process.exit(1);
}

chai.use(chaiPlugins);

function supportsMultipart() {
  return 'SUPPORTS_MULTIPART' in process.env && process.env.SUPPORTS_MULTIPART === 'true';
}

function isListening(childProcess, port, attempt = 0) {
  return new Promise((resolve, reject) => {
    if (childProcess.exitCode !== null) {
      throw new Error(`Unexpected exit code: ${childProcess.exitCode} from child process`);
    }
    if (attempt > 5) throw new Error(`Cannot connect on port: ${port}`);
    const socket = net.connect(port, 'localhost');
    socket.once('error', err => {
      if (err.code !== 'ECONNREFUSED') {
        throw err;
      }
      return setTimeout(() => {
        return isListening(childProcess, port, attempt + 1).then(resolve, reject);
      }, 300 * attempt);
    });

    socket.once('connect', () => {
      return resolve();
    });
  });
}

async function getBody(response) {
  let responseBody = '';
  // eslint-disable-next-line no-restricted-syntax
  for await (const chunk of response) {
    responseBody += chunk;
  }
  expect(responseBody).not.to.equal('');
  return JSON.parse(responseBody);
}

const randomApiKey = 'a-random-readme-api-key';

describe('Metrics SDK Integration Tests', function () {
  let metricsServer;
  let httpServer;
  let PORT;

  before(async function () {
    metricsServer = http.createServer().listen(0, 'localhost');

    await once(metricsServer, 'listening');
    const { address, port } = metricsServer.address();
    PORT = await getPort();

    // In order to use child_process.spawn, we have to provide a
    // command along with an array of arguments. So this is a very
    // rudimental way of splitting the two values provided to us
    // from the environment variable.
    //
    // I tried refactoring this to use child_process.exec, which just
    // takes in a single string to run, but that creates it's own
    // shell so we can't do `cp.kill()` on it later on (because that
    // just kills the shell, not the actual command we're running).
    //
    // Annoyingly this works under macOS, so it must be a platform
    // difference when running under docker/linux.
    const [command, ...args] = process.env.EXAMPLE_SERVER.split(' ');
    if (command === 'php') {
      // Laravel's `artisan serve` command doesn't pick up `PORT` environmental variables, instead
      // requiring that they're supplied as a command line argument.
      args.push(`--port=${PORT}`);
    }

    httpServer = spawn(command, args, {
      cwd: cwd(),
      detached: true,
      env: {
        PORT,
        METRICS_SERVER: new URL(`http://${address}:${port}`).toString(),
        README_API_KEY: randomApiKey,
        ...process.env,
      },
    });

    function prefixStream(prefix) {
      return new Transform({
        transform(chunk, encoding, cb) {
          return cb(
            null,
            chunk
              .toString()
              .split('\n')
              .map(line => `[${prefix}]: ${line}`)
              .join('\n')
          );
        },
      });
    }
    if (process.env.DEBUG) {
      httpServer.stdout.pipe(prefixStream('stdout')).pipe(process.stdout);
      httpServer.stderr.pipe(prefixStream('stderr')).pipe(process.stderr);
    }

    return isListening(httpServer, PORT);
  });

  after(function () {
    /**
     * Instead of running `httpServer.kill()` we need to dust the process group that was created
     * because some languages and frameworks (like Laravel's Artisan server) fire off a sub-process
     * that doesn't get normally cleaned up when we kill the original `php artisan serve` process.
     *
     * Checking that `exitCode` is null before killing group to ensure it is still running
     *
     * @see {@link https://stackoverflow.com/questions/56016550/node-js-cannot-kill-process-executed-with-child-process-exec/56016815#56016815}
     * @see {@link https://www.baeldung.com/linux/kill-members-process-group#killing-a-process-using-the-pgid}
     * @see {@link https://nodejs.org/docs/latest/api/child_process.html#subprocessexitcode}
     */
    if (httpServer.exitCode === null) process.kill(-httpServer.pid);

    return new Promise((resolve, reject) => {
      metricsServer.close(err => {
        if (err) return reject(err);
        return resolve();
      });
    });
  });

  it('should make a request to a Metrics backend with a HAR file', async function () {
    await fetch(`http://localhost:${PORT}`, { method: 'get' });

    const [req, res] = await once(metricsServer, 'request');
    expect(req.url).to.equal('/v1/request');
    expect(req.headers.authorization).to.equal('Basic YS1yYW5kb20tcmVhZG1lLWFwaS1rZXk6');

    const body = await getBody(req);
    const [har] = body;

    // https://uibakery.io/regex-library/uuid
    expect(har._id).to.match(/^[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}$/i);

    expect(har.group).to.deep.equal({
      id: 'owlbert-api-key',
      label: 'Owlbert',
      email: 'owlbert@example.com',
    });

    expect(har.clientIPAddress).to.equal('127.0.0.1');
    expect(har.development).to.be.false;

    const { creator } = har.request.log;
    expect(creator.name).to.match(/readme-metrics \((dotnet|node|php|python|ruby)\)/);
    expect(creator.version).not.to.be.empty;
    expect(creator.comment).not.to.be.empty;

    const { request, response, startedDateTime } = har.request.log.entries[0];

    /**
     * `startedDateTime` should look like the following, with optional microseconds component:
     *
     *  JavaScript: `new Date.toISOString()`
     *    - 2022-06-30T10:21:55.394Z
     *  PHP: `date('Y-m-d\TH:i:sp')`
     *    - 2022-08-17T19:23:31Z
     *  Python: `datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")`
     *    - 2022-06-30T10:31:43Z
     */
    expect(startedDateTime).to.match(/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:.\d{3})?Z/);

    // Some frameworks remove the trailing slash from the URL we get.
    expect(request.url).to.match(new RegExp(`http://localhost:${PORT}(/)?`));
    expect(request.method).to.equal('GET');
    expect(request.httpVersion).to.equal('HTTP/1.1');

    expect(request.headers).to.have.header('connection', [
      'close',
      'keep-alive', // Running this suite with Node 18 the `connection` header is different.
    ]);
    expect(request.headers).to.have.header('host', `localhost:${PORT}`);

    expect(response.status).to.equal(200);
    expect(response.statusText).to.equal('OK');
    expect(response.headers).to.have.header('content-type', /application\/json(;\s?charset=utf-8)?/);

    // Flask prints a \n character after the JSON response
    // https://github.com/pallets/flask/issues/4635
    expect(response.content.text.replace('\n', '')).to.equal(JSON.stringify({ message: 'hello world' }));
    expect(response.content.size).to.equal(response.content.text.length);
    expect(response.content.mimeType).to.match(/application\/json(;\s?charset=utf-8)?/);

    res.end();
    return once(res, 'finish');
  });

  it('should capture query strings in a GET request', async function () {
    await fetch(`http://localhost:${PORT}?arr%5B1%5D=3&val=1`, { method: 'get' });

    const [req, res] = await once(metricsServer, 'request');
    const body = await getBody(req);
    const [har] = body;

    const { request } = har.request.log.entries[0];

    // Some frameworks remove the trailing slash from the URL we get.
    expect(request.url).to.match(new RegExp(`http://localhost:${PORT}(/)?\\?arr%5B1%5D=3&val=1`));

    // Some frameworks handle query string arrays slightly differently.
    expect(JSON.stringify(request.queryString)).to.be.oneOf([
      JSON.stringify([
        { name: 'arr[1]', value: '3' },
        { name: 'val', value: '1' },
      ]),
      JSON.stringify([
        { name: 'arr', value: '{"1":"3"}' },
        { name: 'val', value: '1' },
      ]),
    ]);

    expect(request.postData).to.be.undefined;

    res.end();
    return once(res, 'finish');
  });

  it('should capture query strings that may be supplied in a POST request', async function () {
    const payload = JSON.stringify({ user: { email: 'dom@readme.io' } });
    await fetch(`http://localhost:${PORT}/?arr%5B1%5D=3&val=1`, {
      method: 'post',
      headers: {
        'content-type': 'application/json',
      },
      body: payload,
    });

    const [req, res] = await once(metricsServer, 'request');
    const body = await getBody(req);
    const [har] = body;

    const { request, response } = har.request.log.entries[0];

    expect(request.method).to.equal('POST');

    // Some frameworks remove the trailing slash from the URL we get.
    expect(request.url).to.match(new RegExp(`http://localhost:${PORT}(/)?\\?arr%5B1%5D=3&val=1`));

    expect(request.headers).to.have.header('content-type', 'application/json');

    // Some frameworks handle query string arrays slightly differently.
    expect(JSON.stringify(request.queryString)).to.be.oneOf([
      JSON.stringify([
        { name: 'arr[1]', value: '3' },
        { name: 'val', value: '1' },
      ]),
      JSON.stringify([
        { name: 'arr', value: '{"1":"3"}' },
        { name: 'val', value: '1' },
      ]),
    ]);

    expect(request.postData).to.deep.equal({
      mimeType: 'application/json',
      text: payload,
    });

    expect(response.status).to.equal(200);

    res.end();
    return once(res, 'finish');
  });

  it('should process a POST payload with no explicit `Content-Type` header', async function () {
    const payload = JSON.stringify({ user: { email: 'dom@readme.io' } });
    await fetch(`http://localhost:${PORT}/`, {
      method: 'post',
      body: payload,
    });

    const [req, res] = await once(metricsServer, 'request');
    const body = await getBody(req);
    const [har] = body;

    const { request } = har.request.log.entries[0];

    expect(request.method).to.equal('POST');
    expect(request.headers).to.have.header('content-type', 'text/plain;charset=UTF-8');

    expect(request.postData.mimeType).to.match(/text\/plain(;charset=UTF-8)?/);
    expect(request.postData.params).to.be.undefined;
    expect(request.postData.text).to.equal(payload);

    res.end();
    return once(res, 'finish');
  });

  it('should process an `application/json` POST payload', async function () {
    const payload = JSON.stringify({ user: { email: 'dom@readme.io' } });
    await fetch(`http://localhost:${PORT}/`, {
      method: 'post',
      headers: {
        'content-type': 'application/json',
      },
      body: payload,
    });

    const [req, res] = await once(metricsServer, 'request');
    const body = await getBody(req);
    const [har] = body;

    const { request, response } = har.request.log.entries[0];

    expect(request.method).to.equal('POST');
    expect(request.headers).to.have.header('content-type', 'application/json');
    expect(request.postData).to.deep.equal({
      mimeType: 'application/json',
      text: payload,
    });

    expect(response.status).to.equal(200);

    res.end();
    return once(res, 'finish');
  });

  /**
   * We should eventually support returning the raw POST payload to Metrics in this case but Express
   * has a fun quirk where if you declare the `express.json()` middleware on a route to identify
   * that that route accepts a JSON payload, if that JSON payload is corrupted then it completely
   * wipes out `req.body` and replaces it with an empty JSON object -- eliminating all access for
   * us to the what the original payload was.
   */
  // eslint-disable-next-line mocha/no-pending-tests, mocha/no-skipped-tests
  it.skip('should process an `application/JSON POST payload containing unparseable JSON');

  it('should process a vendored `+json` POST payload', async function () {
    const payload = JSON.stringify({ user: { email: 'dom@readme.io' } });
    await fetch(`http://localhost:${PORT}/`, {
      method: 'post',
      headers: {
        'content-type': 'application/vnd.api+json',
      },
      body: payload,
    });

    const [req, res] = await once(metricsServer, 'request');
    const body = await getBody(req);
    const [har] = body;

    const { request, response } = har.request.log.entries[0];

    expect(request.method).to.equal('POST');
    expect(request.headers).to.have.header('content-type', 'application/vnd.api+json');
    expect(request.postData).to.deep.equal({
      mimeType: 'application/vnd.api+json',
      text: payload,
    });

    expect(response.status).to.be.oneOf([
      200,
      // Fastify doesn't support vendored JSON content types out of the box and will return a
      // `FST_ERR_CTP_INVALID_MEDIA_TYPE` error but thankfully we're still able to capture and
      // process the payload into Metrics.
      415,
    ]);

    res.end();
    return once(res, 'finish');
  });

  it('should process an `application/x-www-url-formencoded` POST payload', async function () {
    const payload = new URLSearchParams();
    payload.append('email', 'dom@readme.io');

    await fetch(`http://localhost:${PORT}/`, {
      method: 'post',
      headers: {
        'content-type': 'application/x-www-form-urlencoded',
      },
      body: payload,
    });

    const [req, res] = await once(metricsServer, 'request');
    const body = await getBody(req);
    const [har] = body;

    const { request, response } = har.request.log.entries[0];

    expect(request.method).to.equal('POST');
    expect(request.headers).to.have.header('content-type', 'application/x-www-form-urlencoded');
    expect(request.postData).to.deep.equal({
      mimeType: 'application/x-www-form-urlencoded',
      params: [{ name: 'email', value: 'dom@readme.io' }],
    });

    expect(response.status).to.be.oneOf([
      200,
      // Fastify, without the `@fastify/formbody` package out of the box doesn't support
      // `x-www-form-urlencoded` and will return a `FST_ERR_CTP_INVALID_MEDIA_TYPE` error.
      // Thankfully our middleware is still able to capture the payload from the request and send
      // that to Metrics regardless if Fastify supports it or not.
      415,
    ]);

    res.end();
    return once(res, 'finish');
  });

  it('should process a `multipart/form-data` POST payload', async function () {
    if (!supportsMultipart()) {
      this.skip();
    }

    const formData = new FormData();
    formData.append('password', 123456);
    formData.append('apiKey', 'abcdef');
    formData.append('another', 'Hello world');
    formData.append('buster', [1234, 5678]);

    const encoder = new FormDataEncoder(formData);

    await fetch(`http://localhost:${PORT}/`, {
      method: 'post',
      headers: encoder.headers,
      body: Readable.from(encoder),
    });

    const [req, res] = await once(metricsServer, 'request');
    const body = await getBody(req);
    const [har] = body;

    const { request } = har.request.log.entries[0];

    expect(request.method).to.equal('POST');
    expect(request.headers).to.have.header('content-type', /multipart\/form-data; boundary=(.*)/);
    expect(request.postData.mimeType).to.match(/multipart\/form-data; boundary=(.*)/);
    expect(request.postData.params).to.deep.equal([
      { name: 'password', value: '123456' },
      { name: 'apiKey', value: 'abcdef' },
      { name: 'another', value: 'Hello world' },
      { name: 'buster', value: '1234,5678' },
    ]);

    expect(request.postData.text).to.be.undefined;

    res.end();
    return once(res, 'finish');
  });

  it('should process a `multipart/form-data` POST payload containing files', async function () {
    if (!supportsMultipart()) {
      this.skip();
    }

    const owlbert = await fs.readFile('./__tests__/__datasets__/owlbert.png');

    const payload = new FormData();
    payload.append('password', 123456);
    payload.append('apiKey', 'abcdef');
    payload.append('another', 'Hello world');
    payload.append('buster', [1234, 5678]);
    payload.append('owlbert.png', new File([owlbert], 'owlbert.png', { type: 'image/png' }), 'owlbert.png');

    const encoder = new FormDataEncoder(payload);

    await fetch(`http://localhost:${PORT}/`, {
      method: 'post',
      headers: encoder.headers,
      body: Readable.from(encoder),
    });

    const [req, res] = await once(metricsServer, 'request');
    const body = await getBody(req);
    const [har] = body;

    const { request } = har.request.log.entries[0];

    expect(request.method).to.equal('POST');
    expect(request.headers).to.have.header('content-type', /multipart\/form-data; boundary=(.*)/);
    expect(request.headers).to.have.header('content-length', 982);
    expect(request.postData.mimeType).to.match(/multipart\/form-data; boundary=(.*)/);

    const owlbertDataURL = await fs.readFile('./__tests__/__datasets__/owlbert.dataurl.json').then(JSON.parse);
    expect(request.postData.params).to.deep.equal([
      { name: 'password', value: '123456' },
      { name: 'apiKey', value: 'abcdef' },
      { name: 'another', value: 'Hello world' },
      { name: 'buster', value: '1234,5678' },
      {
        name: 'owlbert_png',
        value: owlbertDataURL,
        fileName: 'owlbert.png',
        contentType: 'image/png',
      },
    ]);

    expect(request.postData.text).to.be.undefined;

    res.end();
    return once(res, 'finish');
  });

  it('should process a `text/plain` payload', async function () {
    await fetch(`http://localhost:${PORT}/`, {
      method: 'post',
      headers: {
        'content-type': 'text/plain',
      },
      body: 'Hello world',
    });

    const [req, res] = await once(metricsServer, 'request');
    const body = await getBody(req);
    const [har] = body;

    const { request, response } = har.request.log.entries[0];

    expect(request.method).to.equal('POST');
    expect(request.headers).to.have.header('content-type', 'text/plain');
    expect(request.postData).to.deep.equal({
      mimeType: 'text/plain',
      text: 'Hello world',
    });

    expect(response.status).to.equal(200);

    res.end();
    return once(res, 'finish');
  });
});
