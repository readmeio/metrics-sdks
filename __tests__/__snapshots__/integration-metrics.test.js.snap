// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metrics SDK Integration Tests should make a request to a metrics backend with a har file 1`] = `
Object {
  "email": "owlbert@example.com",
  "id": "owlbert-api-key",
  "label": "Owlbert",
}
`;
